Userland/Libraries/LibGfx/TGALoader.cpp

```diff
@@ -0,0 +1,262 @@
+    size_t m_index { 0 };
+};
+
+struct TGALoadingContext {
+    TGAHeader header;
+    ReadonlyBytes bytes;
+    size_t file_size;
+    OwnPtr<TGAReader> reader = { nullptr };
+    RefPtr<Gfx::Bitmap> bitmap;
+};
+
+TGAImageDecoderPlugin::TGAImageDecoderPlugin(u8 const* file_data, size_t file_size)
+{
+    m_context = make<TGALoadingContext>();
+    m_context->bytes = ReadonlyBytes(file_data, file_size);
+    m_context->file_size = move(file_size);
+    m_context->reader = make<TGAReader>(m_context->bytes);
+}
+
+TGAImageDecoderPlugin::~TGAImageDecoderPlugin() = default;
+
+IntSize TGAImageDecoderPlugin::size()
+{
+    return IntSize { m_context->header.width, m_context->header.height };
+}
+
+void TGAImageDecoderPlugin::set_volatile()
+{
+    if (m_context->bitmap)
+        m_context->bitmap->set_volatile();
+}
+
+bool TGAImageDecoderPlugin::set_nonvolatile(bool& was_purged)
+{
+    if (!m_context->bitmap)
+        return false;
+    return m_context->bitmap->set_nonvolatile(was_purged);
+}
+
+bool TGAImageDecoderPlugin::decode_tga_header()
+{
+    auto& reader = m_context->reader;
+    m_context->header = TGAHeader();
+    m_context->header.id_length = reader->read_u8();
+    m_context->header.color_map_type = reader->read_u8();
+    m_context->header.data_type_code = static_cast<TGADataType>(reader->read_u8());
+    m_context->header.color_map_origin = reader->read_i16();
+    m_context->header.color_map_length = reader->read_i16();
+    m_context->header.color_map_depth = reader->read_u8();
+    m_context->header.x_origin = reader->read_i16();
+    m_context->header.y_origin = reader->read_i16();
+    m_context->header.width = reader->read_u16();
+    m_context->header.height = reader->read_u16();
+    m_context->header.bits_per_pixel = reader->read_u8();
+    m_context->header.image_descriptor = reader->read_u8();
+
+    auto bytes_remaining = reader->data().size() - reader->index();
+
+    if (bytes_remaining < (m_context->header.width * m_context->header.height * (m_context->header.bits_per_pixel / 8)))
+        return false;
+
+    if (m_context->header.bits_per_pixel < 8 || m_context->header.bits_per_pixel > 32)
+        return false;
+
+    return true;
+}
+
+bool TGAImageDecoderPlugin::sniff()
+{
+    return decode_tga_header();
+}
+
+bool TGAImageDecoderPlugin::is_animated()
+{
+    return false;
+}
+
+size_t TGAImageDecoderPlugin::loop_count()
+{
+    return 0;
+}
+
+size_t TGAImageDecoderPlugin::frame_count()
+{
+    return 1;
+}
+
+ErrorOr<ImageFrameDescriptor> TGAImageDecoderPlugin::frame(size_t index)
+{
+    auto bits_per_pixel = m_context->header.bits_per_pixel;
+    auto color_map = m_context->header.color_map_type;
+    auto data_type = m_context->header.data_type_code;
+    auto width = m_context->header.width;
+    auto height = m_context->header.height;
+
+    if (index != 0)
+        return Error::from_string_view(String::formatted("BMPImageDecoderPlugin: frame index must be 0 but found {}", index));
```

Comment:
`Error::from_string_view()` must be passed a literal string, not one that's generated in code. (The string data won't be preserved once we leave `TGAImageDecoderPlugin::frame()`.)

A workaround is to create your own Error type that stores a full String instead of StringView, and then return `ErrorOr<ImageFrameDescriptor, MyCoolErrorType>`.